tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h

```diff
@@ -17,6 +20,77 @@
 
 namespace tachyon::math {
 
+namespace {
+
+// NOTE(GideokKim): The formula for memory usage estimation provided in the
+// document did not match the actual memory allocation, so some of the formula
+// was modified. |scalars_memory_size| and |points_memory_size| are exactly the
+// same, and |scalar_indices_memory_size| internally uses the sort function of
+// the cub library to set some free memory. In the case of |bucket_memory_size|,
+// it uses more memory than the actual formula, so it was modified to an
+// empirically more appropriate formula. See
+// https://dev.ingonyama.com/icicle/primitives/msm#memory-usage-estimation
+size_t DetermineMsmDivisionsForMemory(size_t scalar_t_mem_size,
+                                      size_t affine_t_mem_size,
+                                      size_t projective_t_mem_size,
+                                      size_t msm_size, size_t user_c,
+                                      size_t bitsize, size_t precompute_factor,
+                                      size_t batch_size) {
+  size_t free_memory =
+      device::gpu::GpuMemLimitInfo(device::gpu::MemoryUsage::kHigh);
+  size_t shift = 0;
+  size_t log_msm_size = base::bits::Log2Ceiling(msm_size);
+
+  for (size_t number_of_divisions = 0; number_of_divisions < log_msm_size;
+       ++number_of_divisions) {
+    // See
+    // https://github.com/ingonyama-zk/icicle/blob/0cb0b49b/icicle/src/msm/msm.cu#L429-L431
+    size_t c = (user_c == 0) ? static_cast<size_t>(std::max(
+                                   base::bits::Log2Ceiling(msm_size) - 4, 1))
+                             : user_c;
```

Comment:
What does `c` represent here? Can we give more meaningful name?

Reply:
`MSMConfig` also uses `c`, so I used it as is. I think I should just add the part with an explanation of `c` as a comment. What do you think?